src/Cli/dotnet/ToolPackage/ToolPackageDownloader.cs

```diff
@@ -0,0 +1,290 @@
+        private static void AddToolsAssets(
+            ManagedCodeConventions managedCodeConventions,
+            LockFileTargetLibrary lockFileLib,
+            ContentItemCollection contentItems,
+            IReadOnlyList<SelectionCriteria> orderedCriteria)
+        {
+            var toolsGroup = GetLockFileItems(
+                orderedCriteria,
+                contentItems,
+                managedCodeConventions.Patterns.ToolsAssemblies);
+
+            lockFileLib.ToolsAssemblies.AddRange(toolsGroup);
+        }
+
+        private static IEnumerable<LockFileItem> GetLockFileItems(
+            IReadOnlyList<SelectionCriteria> criteria,
+            ContentItemCollection items,
+            params PatternSet[] patterns)
+        {
+            return GetLockFileItems(criteria, items, additionalAction: null, patterns);
+        }
+
+        private static IEnumerable<LockFileItem> GetLockFileItems(
+           IReadOnlyList<SelectionCriteria> criteria,
+           ContentItemCollection items,
+           Action<LockFileItem> additionalAction,
+           params PatternSet[] patterns)
+        {
+            // Loop through each criteria taking the first one that matches one or more items.
+            foreach (var managedCriteria in criteria)
+            {
+                var group = items.FindBestItemGroup(
+                    managedCriteria,
+                    patterns);
+
+                if (group != null)
+                {
+                    foreach (var item in group.Items)
+                    {
+                        var newItem = new LockFileItem(item.Path);
+                        object locale;
+                        if (item.Properties.TryGetValue("locale", out locale))
+                        {
+                            newItem.Properties["locale"] = (string)locale;
+                        }
+                        object related;
+                        if (item.Properties.TryGetValue("related", out related))
+                        {
+                            newItem.Properties["related"] = (string)related;
+                        }
+                        additionalAction?.Invoke(newItem);
+                        yield return newItem;
+                    }
+                    // Take only the first group that has items
+                    break;
+                }
+            }
+
+            yield break;
+        }
+
+        private static async Task<NuGetVersion> DownloadAndExtractPackage(
+            PackageLocation packageLocation,
+            PackageId packageId,
+            INuGetPackageDownloader _nugetPackageDownloader,
+            string hashPathLocation
+            )
+        {
+            var packageSourceLocation = new PackageSourceLocation(packageLocation.NugetConfig, packageLocation.RootConfigDirectory, null, packageLocation.AdditionalFeeds);
+            var packagePath = await _nugetPackageDownloader.DownloadPackageAsync(packageId, null, packageSourceLocation);
```

Comment:
I see that the downloading is hidden deep in this package downloader. 

https://github.com/JL03-Yue/sdk/blob/e6ce77b8ea041a5a775a850b8ee5325e7f60626b/src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs#L313.

When dealing with a settings object, you can create a source mapping object. 

https://github.com/NuGet/NuGet.Client/blob/4d815bf71aa1ce5fe8d9d9da6573303d7093c2f1/src/NuGet.Core/NuGet.Configuration/PackageSourceMapping/PackageSourceMapping.cs#L66

Then anytime *before* you access a source, you'd check https://github.com/NuGet/NuGet.Client/blob/4d815bf71aa1ce5fe8d9d9da6573303d7093c2f1/src/NuGet.Core/NuGet.Configuration/PackageSourceMapping/PackageSourceMapping.cs#L38. 

PSM promises that there's no calls made to unmapped sources, so looking up a version on an unmapped source is not supposed to be done. 

https://github.com/NuGet/NuGet.Client/blob/4d815bf71aa1ce5fe8d9d9da6573303d7093c2f1/src/NuGet.Core/NuGet.DependencyResolver.Core/ResolverUtility.cs#L49

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteWalkContext.cs#L61

Restore itself also does lots of logging when PSM is used and when it is not. 
For example, when a source is *not* used, that's useful debugging information.